providers/implementations/keymgmt/rsa_kmgmt.c

```diff
@@ -96,88 +104,142 @@ static int params_to_key(RSA *rsa, const OSSL_PARAM params[])
+            param_qinv = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_QINV);
+            if ((param_dp != NULL && !OSSL_PARAM_get_BN(param_dp, &dp))
+                || (param_dq != NULL && !OSSL_PARAM_get_BN(param_dq, &dq))
+                || (param_qinv != NULL && !OSSL_PARAM_get_BN(param_qinv, &qinv))
+                || !RSA_set0_crt_params(rsa, dp, dq, qinv))
+                goto err;
+            dp = dq = qinv = NULL;
+#ifndef FIPS_MODE
+            if (!collect_numbers(factors = sk_BIGNUM_new_null(), params,
+                                 OSSL_PKEY_PARAM_RSA_MP_FACTOR)
+                || !collect_numbers(exps = sk_BIGNUM_new_null(), params,
+                                    OSSL_PKEY_PARAM_RSA_MP_EXPONENT)
+                || !collect_numbers(coeffs = sk_BIGNUM_new_null(), params,
+                                    OSSL_PKEY_PARAM_RSA_MP_COEFFICIENT))
+                goto err;
+
+            /* It's ok if this private key just has n, e and d */
+            if (sk_BIGNUM_num(factors) != 0
+                && !rsa_set0_all_mp_params(rsa, factors, exps, coeffs))
+                goto err;
+#endif
+        }
     }
 
+#ifndef FIPS_MODE
     sk_BIGNUM_free(factors);
     sk_BIGNUM_free(exps);
     sk_BIGNUM_free(coeffs);
+#endif
     return 1;
 
  err:
     BN_free(n);
     BN_free(e);
     BN_free(d);
+    BN_free(p);
+    BN_free(q);
+    BN_free(dp);
+    BN_free(dq);
+    BN_free(qinv);
+#ifndef FIPS_MODE
     sk_BIGNUM_pop_free(factors, BN_free);
     sk_BIGNUM_pop_free(exps, BN_free);
     sk_BIGNUM_pop_free(coeffs, BN_free);
+#endif
     return 0;
 }
 
-static int export_numbers(OSSL_PARAM_BLD *tmpl, const char *key,
-                          STACK_OF(BIGNUM_const) *numbers)
+static int set_bn(OSSL_PARAM_BLD *bld, OSSL_PARAM *p, const char *key,
+                  const BIGNUM *bn)
 {
-    int i, nnum;
+    if (bld != NULL)
+        return ossl_param_bld_push_BN(bld, key, bn);
 
-    if (numbers == NULL)
-        return 0;
+    p = OSSL_PARAM_locate(p, key);
+    if (p != NULL)
+        return OSSL_PARAM_set_BN(p, bn) > 0;
+    return 1;
+}
 
-    nnum = sk_BIGNUM_const_num(numbers);
+#ifndef FIPS_MODE
+static int set_multi_key_bn(OSSL_PARAM_BLD *bld, OSSL_PARAM *p,
+                            const char *key, STACK_OF(BIGNUM_const) *stk)
+{
+    int i, sz = sk_BIGNUM_const_num(stk);
+
+    if (bld != NULL) {
 
-    for (i = 0; i < nnum; i++) {
-        if (!ossl_param_bld_push_BN(tmpl, key,
-                                    sk_BIGNUM_const_value(numbers, i)))
+        for (i = 0; i < sz; ++i) {
+            if (!ossl_param_bld_push_BN(bld, key, sk_BIGNUM_const_value(stk, i)))
+                return 0;
+        }
+        return 1;
+    }
+    p = OSSL_PARAM_locate(p, key);
+    for (i = 0; p != NULL && i < sz; ++i) {
+        if (!OSSL_PARAM_set_BN(p, sk_BIGNUM_const_value(stk, i)))
             return 0;
+        p = OSSL_PARAM_locate(p, key);
```

Comment:
`p = OSSL_PARAM_locate(p, key)` will return the same pointer.  May I suggest this instead?

``` C
    for (i = 0; p != NULL && i < sz; ++i, ++p) {
        if ((p = OSSL_PARAM_locate(p, key)) != NULL
            && !OSSL_PARAM_set_BN(p, sk_BIGNUM_const_value(stk, i)))
            return 0;
    }
```

Reply:
Fixed in #11365.